### PR TITLE
RAT-384: write xsd (RAT-384 part 2) (Harmonize - Part 5)

### PR DESCRIPTION
**Pull request by Claudenw**
Creates an XSD for configuration files to verify configuration structure.
Updates documentation for Maven plugin.

### CODE REVIEW

File: apache-rat-core/src/main/java/org/apache/rat/help/AbstractHelp.java [hunk 1]
```diff
@@ -78,4 +80,98 @@ public static String createPadding(final int len) {
...
+                helpFormatter.printWrapped(pw, HELP_WIDTH, header);
+            }
+            printOptions(pw, HELP_WIDTH, options, helpFormatter.getLeftPadding(), helpFormatter.getDescPadding());
+            if (footer != null && !footer.isEmpty()) {
+                helpFormatter.printWrapped(pw, helpFormatter.getWidth(), footer);
+            }
+        }
+
+
+        protected StringBuffer renderOptions(final StringBuffer sb, final int width, final Options options, final int leftPad, final int descPad) {
```
Comment (ottlinger) [inline apache-rat-core/src/main/java/org/apache/rat/help/AbstractHelp.java:109]: Isn't StringBuilder due to performance issues a better choice here?

File: apache-rat-core/src/main/java/org/apache/rat/report/xml/writer/impl/base/XmlWriter.java [hunk 2]
```diff
@@ -627,6 +627,39 @@ public IXmlWriter closeElement() throws IOException {
...
 
+    /**
+     * Closes back to and including the last instance of the specified element nema.
+     * @param name The name of the element to close.
+     * @return this object
+     * @throws OperationNotAllowedException if called before any call to
+     * {@link #openElement} or after the first element has been closed
+     */
+    @Override
+    public IXmlWriter closeElement(CharSequence name) throws IOException {
```
Comment (ottlinger) [inline apache-rat-core/src/main/java/org/apache/rat/report/xml/writer/impl/base/XmlWriter.java:638]: Is NPE-safeness an issue here? Should we throw an exception if invalid/null is passed in here?

File: apache-rat-tools/src/main/java/org/apache/rat/tools/xsd/XsdWriter.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    public XsdWriter init() throws IOException {
+        writer.startDocument()
+        .openElement("xs:schema")
+        .attribute("attributeFormDefault", "unqualified")
+                .attribute("xmlns:xs", "http://www.w3.org/2001/XMLSchema");
+        return this;
+    }
+
+    public void fini() throws IOException {
```
Comment (ottlinger) [inline apache-rat-tools/src/main/java/org/apache/rat/tools/xsd/XsdWriter.java:54]: should this be finish() instead?

File: apache-rat-tools/src/test/java/org/apache/rat/tools/NamingTest.java [hunk 4]
```diff
@@ -181,10 +181,4 @@ private void assertContains(String expected, List<String> actual) {
     private void assertNotContains(String expected, List<String> actual) {
         assertFalse(actual.contains(expected), () -> "Contains "+expected);
     }
-
-    @Test
```
Comment (ottlinger) [inline apache-rat-tools/src/test/java/org/apache/rat/tools/NamingTest.java:185]: If I'm not too mistaken I added these tests to validate the NPE-safeness of the Naming class? Is that not needed anymore?

File: src/site/apt/license_def.apt.vm [hunk 5]
```diff
@@ -29,7 +29,9 @@ How to define licenses in Apache Rat
  scope of this document.  In this document we will address how to define licenses in the default XML format.
  
  The XML document has a root node named "rat-config" which comprises 4 major sections: "Families", 
- "Licenses", "Approved" and "Matchers".
+ "Licenses", "Approved" and "Matchers".  There is an {{{./apache-rat/rat.xsd}}XSD} that can be used to validate
```
Comment (ottlinger) [inline src/site/apt/license_def.apt.vm:32]: @Claudenw is there a technical reason for each sentence to start with 2 spaces? (or is it just a formatter setting of yours)

File: src/site/apt/index.apt.vm [hunk 6]
```diff
@@ -183,7 +177,7 @@ cd creadur-rat
 mvn clean install
 +------------------------------------------+
 
-or with the included Maven wrapper
+ or with the included Maven wrapper
```
Comment (ottlinger) [inline src/site/apt/index.apt.vm:180]: I assume this is already fixed in master ....

**Comment by ottlinger** [comment]
@Claudenw WDYT - should we add a changelog entry for this one as it enhances the functionality and introduces the whole XSD/configuration handling?

File: src/site/apt/license_def.apt.vm (hunk 5 above)
Comment (Claudenw) [reply src/site/apt/license_def.apt.vm]: just a formatter setting.  I think I had a problem at one point and just ended up using 2 spaces.

File: apache-rat-tools/src/test/java/org/apache/rat/tools/NamingTest.java (hunk 4 above)
Comment (Claudenw) [reply apache-rat-tools/src/test/java/org/apache/rat/tools/NamingTest.java:185]: Missed it in a merge somehow.

File: apache-rat-tools/src/main/java/org/apache/rat/tools/xsd/XsdWriter.java (hunk 3 above)
Comment (Claudenw) [reply apache-rat-tools/src/main/java/org/apache/rat/tools/xsd/XsdWriter.java:54]: yes

File: apache-rat-core/src/main/java/org/apache/rat/report/xml/writer/impl/base/XmlWriter.java (hunk 2 above)
Comment (Claudenw) [reply apache-rat-core/src/main/java/org/apache/rat/report/xml/writer/impl/base/XmlWriter.java:637]: added Object.requireNonNull and updated javadoc to say nulls not allowed.

**Comment by Claudenw** [comment]
This change has exhibited a weird error.  It seem like it is pulling in commons-cli 1.9.0 while everything in the codebase seems to point to commons-cli 1.8.0

**Comment by Claudenw** [comment]
@ottlinger if you approve of this we can merge it now.

The CLI update will have to wait for commons-cli 1.10.0

**Comment by ottlinger** [comment]
@Claudenw did you push all changes? The current version only uses v1.8.0:

```
➜  creadur-rat git:(RAT-384_write_XSD) ./mvnw dependency:tree
.....
[INFO] 
[INFO] ------------------< org.apache.rat:apache-rat-tools >-------------------
[INFO] Building Apache Creadur Rat::Tools 0.17-SNAPSHOT                   [3/6]
[INFO]   from apache-rat-tools/pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- dependency:3.8.0:tree (default-cli) @ apache-rat-tools ---
[INFO] org.apache.rat:apache-rat-tools:jar:0.17-SNAPSHOT
[INFO] +- org.apache.rat:apache-rat-core:jar:0.17-SNAPSHOT:compile
[INFO] |  +- org.apache.commons:commons-collections4:jar:4.4:compile
[INFO] |  +- org.apache.commons:commons-lang3:jar:3.17.0:compile
[INFO] |  +- commons-io:commons-io:jar:2.16.1:compile
[INFO] |  +- org.apache.commons:commons-text:jar:1.12.0:compile
[INFO] |  +- org.junit.jupiter:junit-jupiter-api:jar:5.11.0:compile
[INFO] |  |  +- org.opentest4j:opentest4j:jar:1.3.0:compile
[INFO] |  |  +- org.junit.platform:junit-platform-commons:jar:1.11.0:compile
[INFO] |  |  \- org.apiguardian:apiguardian-api:jar:1.1.2:compile
[INFO] |  +- org.junit.vintage:junit-vintage-engine:jar:5.11.0:compile
[INFO] |  |  +- org.junit.platform:junit-platform-engine:jar:1.11.0:compile
[INFO] |  |  \- junit:junit:jar:4.13.2:compile
[INFO] |  |     \- org.hamcrest:hamcrest-core:jar:1.3:compile
[INFO] |  +- org.junit.jupiter:junit-jupiter-params:jar:5.11.0:compile
[INFO] |  \- org.apache.tika:tika-core:jar:2.9.2:compile
[INFO] |     \- org.slf4j:slf4j-api:jar:2.0.10:compile
[INFO] +- org.apache.rat:apache-rat-core:test-jar:tests:0.17-SNAPSHOT:test
[INFO] +- commons-cli:commons-cli:jar:1.8.0:compile
[INFO] +- org.apache.commons:commons-csv:jar:1.11.0:compile
[INFO] |  \- commons-codec:commons-codec:jar:1.16.1:compile
[INFO] \- org.apache.commons:commons-compress:jar:1.27.1:compile
[INFO] 

```

**Comment by Claudenw** [comment]
@ottlinger 

Yes, the current version uses commons-cli 1.8.0, v1.9.0 changes the help generation in ways that broke our output.  CLI v1.10.0 should fix it but I am still working on it.  I was hoping to get a first cut done this weekend but that it not looking good right now.

File: src/site/apt/index.apt.vm [hunk 7]
```diff
@@ -99,10 +99,6 @@ Apache Rat&#8482;
   There are three user interfaces for Rat: from the command line, from Ant, and from Maven.  All three of the interfaces
   support standard options as noted in the list below.
 
-+------------------------------------------+
-
```
Comment (ottlinger) [inline src/site/apt/index.apt.vm:103]: @Claudenw is there an inclusion of a txt file missing here?

**Comment by ottlinger** [comment]
@Claudenw if you are fine with my changes feel free to merge and go ahead.

**Review by ottlinger — APPROVED** [review]
Minor question concerning the inclusion of a txt file and trivial changes in files.

**Pull request merged**